chainservice/builder.go

```diff
@@ -0,0 +1,689 @@
+	indexer blockindex.Indexer,
+	bfIndexer blockindex.BloomFilterIndexer,
+	candidateIndexer *poll.CandidateIndexer,
+	err error,
+) {
+	_, gateway := builder.cfg.Plugins[config.GatewayPlugin]
+	if !gateway {
+		return
+	}
+
+	if forTest {
+		indexer, err = blockindex.NewIndexer(db.NewMemKVStore(), builder.cfg.Genesis.Hash())
+		if err != nil {
+			return
+		}
+		bfIndexer, err = blockindex.NewBloomfilterIndexer(db.NewMemKVStore(), builder.cfg.Indexer)
+		if err != nil {
+			return
+		}
+		candidateIndexer, err = poll.NewCandidateIndexer(db.NewMemKVStore())
+
+		return
+	}
+	dbConfig := builder.cfg.DB
+	dbConfig.DbPath = builder.cfg.Chain.IndexDBPath
+	indexer, err = blockindex.NewIndexer(db.NewBoltDB(dbConfig), builder.cfg.Genesis.Hash())
+	if err != nil {
+		return
+	}
+
+	// create bloomfilter indexer
+	dbConfig.DbPath = builder.cfg.Chain.BloomfilterIndexDBPath
+	bfIndexer, err = blockindex.NewBloomfilterIndexer(db.NewBoltDB(dbConfig), builder.cfg.Indexer)
+	if err != nil {
+		return
+	}
+
+	// create candidate indexer
+	dbConfig.DbPath = builder.cfg.Chain.CandidateIndexDBPath
+	candidateIndexer, err = poll.NewCandidateIndexer(db.NewBoltDB(dbConfig))
+
+	return
+}
+
+func (builder *Builder) buildBlockchain(forSubChain, forTest bool) error {
+	builder.cs.chain = builder.createBlockchain(forSubChain, forTest)
+	builder.cs.lifecycle.Add(builder.cs.chain)
+
+	if err := builder.cs.chain.AddSubscriber(builder.cs.actpool); err != nil {
+		return errors.Wrap(err, "failed to add actpool as subscriber")
+	}
+	if builder.cs.indexer != nil && builder.cfg.Chain.EnableAsyncIndexWrite {
+		// config asks for a standalone indexer
+		indexBuilder, err := blockindex.NewIndexBuilder(builder.cs.chain.ChainID(), builder.cs.blockdao, builder.cs.indexer)
+		if err != nil {
+			return errors.Wrap(err, "failed to create index builder")
+		}
+		builder.cs.lifecycle.Add(indexBuilder)
+		if err := builder.cs.chain.AddSubscriber(indexBuilder); err != nil {
+			return errors.Wrap(err, "failed to add index builder as subscriber")
+		}
+	}
+	return nil
+}
+
+func (builder *Builder) createBlockchain(forSubChain, forTest bool) blockchain.Blockchain {
+	if builder.cs.chain != nil {
+		return builder.cs.chain
+	}
+	var chainOpts []blockchain.Option
+	if !forSubChain {
+		chainOpts = append(chainOpts, blockchain.BlockValidatorOption(block.NewValidator(builder.cs.factory, builder.cs.actpool)))
+	} else {
+		chainOpts = append(chainOpts, blockchain.BlockValidatorOption(builder.cs.factory))
+	}
+
+	return blockchain.NewBlockchain(builder.cfg, builder.cs.blockdao, factory.NewMinter(builder.cs.factory, builder.cs.actpool), chainOpts...)
+}
+
+func (builder *Builder) buildBlockSyncer() error {
+	if builder.cs.blocksync != nil {
+		return nil
+	}
+	if builder.cfg.Consensus.Scheme == config.StandaloneScheme {
+		builder.cs.blocksync = blocksync.NewDummyBlockSyncer()
+		return nil
+	}
+
+	p2pAgent := builder.cs.p2pAgent
+	chain := builder.cs.chain
+	consens := builder.cs.consensus
```

Comment:
check nil for these 3?

Reply:
They are inside of one struct. If we have a separate builder for it, I will consider to add these checks.

Reply:
it relies on `buildBlockchain` called first, so here `builder.cs.chain` is not nil, for example, the below order would cause problem:
```
if err := builder.buildBlockSyncer(); err != nil {
	return nil, err
}
if err := builder.buildBlockchain(); err != nil {
	return nil, err
}
```
check nil can safeguard this